### PR TITLE
fix: update tracer for llama-index 0.9.0

### LINKED ISSUE #1747: [BUG] phoenix tracing crashing with latest llama_index
From Llama_index

This snippet of the traceback may be helpful:
```
   return func(*args, **kwargs)
  File "/Users/nerdai/Library/Caches/pypoetry/virtualenvs/llama-index-e6cjsBOJ-py3.10/lib/python3.10/site-packages/phoenix/trace/llama_index/callback.py", line 266, in on_event_start
    payload_to_semantic_attributes(event_type, payload),
  File "/Users/nerdai/Library/Caches/pypoetry/virtualenvs/llama-index-e6cjsBOJ-py3.10/lib/python3.10/site-packages/phoenix/trace/llama_index/callback.py", line 194, in payload_to_semantic_attributes
    attributes[TOOL_PARAMETERS] = tool_metadata.to_openai_function()["parameters"]
KeyError: 'parameters' 
```
We've made a recent release that deprecated to_openai_function() to mirror the latest OpenAI API. This should be changed to to_openai_tool but more importantly the parameters key lives now one step down in resulting dictionary:
so, something like this should do the trick
```
tool_metadata.to_openai_tool()["function"]["parameters"]
```

### PR DESCRIPTION
**Pull request by RogerHYang**
resolves #1747 

# Key Changes
1. Added `tool` as a new message role, because the `function` role is deprecated. (This is the result of the function call that we send back to openai as part of a chat)
2. Check the minimum version for llama-index at `0.9.0` at when initializing the tracer.
3. For the `ToolMetata` object from llama-index, change `tool_metadata.to_openai_function()["parameters"]` to `tool_metadata.to_openai_tool()["function"]["parameters"]` because the former is deprecated, and its underlying dictionary structure has changed.
4. Multiple function calls is now possible in the OpenAI response message (i.e. the new `tool_calls` is a list, the old `function_call`, which is singular, is deprecated). But we're only capturing the first one for now. A follow-up PR will be needed to capture all.

### CODE REVIEW

File: src/phoenix/trace/llama_index/callback.py [hunk 1]
```diff
@@ -226,6 +229,15 @@ def __init__(
...
     ) -> None:
+        if (
+            hasattr(llama_index, "__version__")
+            and (version_triplet := extract_version_triplet(llama_index.__version__))
+            and version_triplet < LLAMA_INDEX_MINIMUM_VERSION_TRIPLET
+        ):
+            raise NotImplementedError(
+                f"minimum supported version of llama-index is "
+                f"{'.'.join(map(str, LLAMA_INDEX_MINIMUM_VERSION_TRIPLET))}"
+            )
```
Comment (mikeldking) [inline src/phoenix/trace/llama_index/callback.py:240]: Maybe just advise the user to upgrade?

File: src/phoenix/trace/llama_index/callback.py [hunk 2]
```diff
@@ -437,10 +449,18 @@ def _message_payload_to_attributes(message: Any) -> Dict[str, Optional[str]]:
...
+                tool_calls, Iterable
+            ), f"tool_calls must be Iterable, found {type(tool_calls)}"
+            tool_call = next(iter(tool_calls), None)
+            function = getattr(tool_call, "function", None)
+            if name := getattr(function, "name", None):
+                assert isinstance(name, str), f"name must be str, found {type(name)}"
+                message_attributes[MESSAGE_FUNCTION_CALL_NAME] = name
+            if arguments := getattr(function, "arguments", None):
+                assert isinstance(arguments, str), f"arguments must be str, found {type(arguments)}"
+                message_attributes[MESSAGE_FUNCTION_CALL_ARGUMENTS_JSON] = arguments
```
Comment (mikeldking) [inline src/phoenix/trace/llama_index/callback.py:463]: is this related to the multi-function calling in GPT-4-Turbo? If so we probably need a follow-up ticket. Seems like we are only parsing the first tool call?

**Review by mikeldking — APPROVED** [review]
Left some questions about the Iterable tool calls. Also tool is now a message role? Otherwise looks good to me.

**Comment by mikeldking** [comment]
@RogerHYang can you add some color to the PR content just so it's searchable and grokable? It can be short and sweet.

File: src/phoenix/trace/llama_index/callback.py [hunk 3]
```diff
@@ -191,7 +194,7 @@ def payload_to_semantic_attributes(
         tool_metadata = cast(ToolMetadata, payload.get(EventPayload.TOOL))
         attributes[TOOL_NAME] = tool_metadata.name
         attributes[TOOL_DESCRIPTION] = tool_metadata.description
-        attributes[TOOL_PARAMETERS] = tool_metadata.to_openai_function()["parameters"]
+        attributes[TOOL_PARAMETERS] = tool_metadata.to_openai_tool()["function"]["parameters"]
```
[the remark this answers is not in the document]
Comment (mikeldking) [reply src/phoenix/trace/llama_index/callback.py:197]: nvm, I see below

File: app/src/pages/trace/TracePage.tsx [hunk 4]
```diff
@@ -918,7 +918,7 @@ function LLMMessage({ message }: { message: AttributeMessage }) {
         backgroundColor: "indigo-100",
         borderColor: "indigo-700",
       };
-    } else if (role === "function") {
+    } else if (["function", "tool"].includes(role)) {
```
[the remark this answers is not in the document]
Comment (RogerHYang) [reply app/src/pages/trace/TracePage.tsx:921]: yup, this is the result of the function call that we send back to openai as part of a chat

Screenshot 2023-11-15 at 4 29 39 PM

File: pyproject.toml [hunk 5]
```diff
@@ -110,7 +110,7 @@ dependencies = [
 [tool.hatch.envs.type]
 dependencies = [
   "mypy==1.5.1",
-  "llama-index>=0.8.64",
+  "llama-index>=0.9.0",
```
[the remark this answers is not in the document]
Comment (RogerHYang) [reply pyproject.toml:113]: good idea. will add

File: src/phoenix/trace/llama_index/callback.py [hunk 6]
```diff
@@ -160,9 +162,10 @@ def payload_to_semantic_attributes(
             attributes[INPUT_VALUE] = _message_payload_to_str(messages[0])
     if response := (payload.get(EventPayload.RESPONSE) or payload.get(EventPayload.COMPLETION)):
         attributes.update(_get_response_output(response))
-        if (raw := getattr(response, "raw", None)) is not None:
+        if raw := getattr(response, "raw", None):
+            assert isinstance(raw, Mapping), f"raw must be Mapping, found {type(raw)}"
```
[the remark this answers is not in the document]
Comment (RogerHYang) [reply src/phoenix/trace/llama_index/callback.py:166]: yes, either way it'll crash on the `.get` line below if it ceases to be a mapping, so this is just intended to make debugging slightly more transparent.

File: src/phoenix/trace/llama_index/callback.py (hunk 2 above)
Comment (RogerHYang) [reply src/phoenix/trace/llama_index/callback.py:463]: yea, it complicates everything. i will add a follow-up ticket

File: src/phoenix/trace/llama_index/callback.py (hunk 1 above)
Comment (RogerHYang) [reply src/phoenix/trace/llama_index/callback.py:240]: good idea. will add

**Pull request merged**